### PR TITLE
FileProvider Integration for external file permissions

### LINKED ISSUE #2263: FileUriExposedException when taking a picture for an image question
#### Software and hardware versions 
All Collect versions, Android API 24, Google Nexus 6

#### Problem description
Once the target API is set for API level 24 or greater the app crashes when you try to take a picture i.e when a file created for the image is being passed to the camera app.

` Process: org.odk.collect.android, PID: 4311
    android.os.FileUriExposedException: file:///storage/emulated/0/odk/.cache/tmp.jpg exposed beyond app through ClipData.Item.getUri()
        at android.os.StrictMode.onFileUriExposed(StrictMode.java:1799)
        at android.net.Uri.checkFileUriExposed(Uri.java:2346)`

#### Steps to reproduce the problem
Set the target API version to 24 or higher and attempt to take a picture.
#### Expected behavior
The app should launch the camera app and accept the picture that's taken as an answer.

### PR DESCRIPTION
**Pull request by jd-alexander**
Closes #2263

This PR addresses the `FileUriExposedException` as described in the tagged issue. This PR is a precursor to the addition of camera permissions since this exception is thrown as soon as the camera is accessed on Nougat and later.

The main things that were done to address this are 

- [x] `FileProvider` from the support package was registered within the manifest to allow sharing of files with external apps.
- [x] The Annotate, Arbitrary File and Image Widgets were modified to provide an uri to the target external app that it can write/read. 
- [x] Utility functions were created to grant and revoke permissions of a specified uri. These functions have backward compat support for all versions of Android.

#### What has been done to verify that this works as intended?
Once you set the target SDK to API 24 or higher and you attempt to take a picture this exception is thrown. However when you you switch to this branch/PR and do the same the it bypasses that exception since the `FileProvider` is granting the external app i.e the camera app permission to read/write to the file. A permission denial exception is then thrown because the permission grants haven't been implemented as yet so that's the validation being used since this a WIP PR. 

#### Why is this the best possible solution? Were any other approaches considered?
While testing you can utilize Android's Strict Mode functionality to bypass this but that doesn't suit production so this approach is the best since it utilizes the ContentProvider to provide access to external apps and that's basically a default standard in Android cross app/process communication.

#### Are there any risks to merging this code? If so, what are they?
Not necessarily. We just have to verify that all functionality within the app that are sharing data with external apps are working fine on Nougat and if not then utilize this `FileProvider` approach to do so. I checked the `AboutActivity` that's performing several shares and external links and it seems to be fine; i.e those intents such as `ACTION_SEND` don't require these grants.

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form with image questions.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lint findbugs` and confirmed all checks still pass.
All checks pass once the Api version is below 24. Once it's set to that or above lint complains about the permissions that haven't been implemented.
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by jd-alexander** [comment]
@grzesiek2010 this is what I have done so far in terms of starting the camera permissions. i.e I have adddress the file permissions for the uri being shared with the camera app so the next likely step is to simply merge this branch into a `camera_permissions` branch and then implement the permission grants/checks on top of this.

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,10 +1,6 @@
...
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-android.enableD8.desugaring=true
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.7-bin.zip
-# Android Studio Gradle wrapper 3.1.2 doesn't support configure on demand
-org.gradle.configureondemand=false
\ No newline at end of file
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.7-all.zip
```
Comment (lognaturel) [inline gradle/wrapper/gradle-wrapper.properties]: Please revert all changes to this file. 😊

Comment (jd-alexander) [reply gradle/wrapper/gradle-wrapper.properties]: oh yes! this always slips in.

**Comment by mmarciniak90** [comment]
@jd-alexander 
I'm not sure that I understand everything correct, so I describe what I see:
*  This permission dialog is visible only for selfie image and selfie video screenshot_20180614-123231
* The app crashes when I try to add an image when permission is not added. I expect that I will see the same dialog as mentioned above. If it is implemented in other PR, maybe it is possible to update this PR to contain it.
* The app does not crash when I try to add an image when permission was added manually :+1:

**Comment by jd-alexander** [comment]
Yes all of this is the expected behaviour.

> This permission dialog is visible only for selfie image and selfie video 

Yes, this functionality has runtime permissions implemented.

> The app crashes when I try to add an image when permission is not added. I expect that I will see the same dialog as mentioned above. If it is implemented in other PR, maybe it is possible to update this PR to contain it.

Yes that's what will happen when the functionality is added.

> The app does not crash when I try to add an image when permission was added manually

Yup, to even go a bit further and verify that this PR is targeting the file permissions issue properly you can go to another branch, set the target api version to 24 and when you turn on the image permissions and try to take a picture it should crash.

**Comment by mmarciniak90** [comment]
@jd-alexander thank you for explanations.

**Tested with success**

Verified on Android: 6.0, 7.0, 8.1
Verified also Android 4.2, 4.4, 5.1 and regression was not noticed

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

**Comment by lognaturel** [comment]
Thanks, @jd-alexander, @mmarciniak90! @jd-alexander can you please fix the merge conflicts? Then this is ready for merging.

**Comment by jd-alexander** [comment]
@lognaturel yes gonna address them!

**Comment by jd-alexander** [comment]
@lognaturel ready!

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 2]
```diff
@@ -695,6 +697,14 @@ protected void onActivityResult(int requestCode, int resultCode, final Intent in
                 // file
                 ImageConverter.execute(Collect.TMPFILE_PATH, getWidgetWaitingForBinaryData(), this);
                 File fi = new File(Collect.TMPFILE_PATH);
+
+                //revoke permissions granted to this file due it's possible usage in
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: Oops, hadn't noticed this before. "due it's possible" -> "due to its possible usage"

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 3]
```diff
@@ -505,4 +512,64 @@ public static void write(File file, byte[] data) {
             Timber.e(e);
         }
     }
+
+    /**
+     * With the FileProvider you have to manually grand and revoke read/write permissions to files you
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java]: "grand" -> "grant"

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 4]
```diff
@@ -505,4 +512,64 @@ public static void write(File file, byte[] data) {
...
+     * are sharing. With this approach the access only last as long as the target activity on Api versions
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java]: "only last" -> "only lasts"

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 5]
```diff
@@ -505,4 +512,64 @@ public static void write(File file, byte[] data) {
...
+            for (ResolveInfo resolveInfo : resInfoList) {
+                String packageName = resolveInfo.activityInfo.packageName;
+                context.grantUriPermission(packageName, uri, Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
+            }
+        }
+    }
+
+
+    /**
+     * With the FileProvider you have to manually grand and revoke read/write permissions to files you
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java]: "grand" -> "grant"

File: collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java [hunk 6]
```diff
@@ -505,4 +512,64 @@ public static void write(File file, byte[] data) {
...
+     * are sharing. With this approach the access only last as long as the target activity on Api versions
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/utilities/FileUtils.java]: "only last" -> "only lasts"

**Review by lognaturel — COMMENTED** [review]
Sorry, I had missed some little proofreading issues!

**Review by lognaturel — APPROVED** [review]

**Pull request merged**